lapack/gonum,testlapack: fix bug in Dtrevc3 and rework its test

vladimir-ch opened the pull request:
Please take a look.

<!--
Checklist:

- API changes have been discussed
- code is goformated correctly (goimports)
- packages with generated code have had code generation run
- tests pass locally
- linked to relevant issues

Please make sure your commit message summary line and pull request
title match the Go convention; a one-line summary of the change,
prefixed by the primary affected package that should complete the
sentence, "This change modifies Gonum to _____."
-->

vladimir-ch commented on lapack/gonum/dtrevc3.go:
```diff
@@ -647,7 +647,7 @@ leftev:
 					beta := math.Max(norms[j], norms[j+1])
 					if beta > vcrit {
-						bi.Dscal(n-ki+1, 1/vmax, b[ki*ldb+iv:], 1)
+						bi.Dscal(n-ki, 1/vmax, b[ki*ldb+iv:], ldb)
```
This shows the importance of using also bad matrices for testing. This was started to be covered only after I added zero/tiny eigenvalues to `randomSchurCanonical`.

kortschak commented:
This appears to have flowed on to many `Dtrexc` failures. Is this something that we should be worried about, or should we enable conditional return of zero/tiny eigenvalues from `randomSchurCanonical`?

vladimir-ch commented:
Sorry, I was running only the Dtrevc3 test locally. I had thought about enabling zero/tiny values conditionally but then decided against it for simplicity. I'll take a look today but at least the "unexpected eigenvalue of 2x2 block" failure looks like an issue with the test.

vladimir-ch commented:
In the end I've added a flag to `randomSchurCanonical` because checking Dtrexc is another rabbit hole and I need time for exploring it. I've checked, reworked and (hopefully) simplified the test for Dtrexc but for bad matrices the test consistently fails. For good matrices the coverage is only 50% while with bad matrices it's practically complete, so it's still possible there's a bug.

vladimir-ch commented on lapack/testlapack/dtrexc.go:
```diff
@@ -205,16 +199,126 @@ func testDtrexc(t *testing.T, impl Dtrexcer, compq lapack.UpdateSchurComp, tmat
+	}
+}
+
+func residualOrthogonal(q blas64.General, rowwise bool) float64 {
```
Yes, this is yet another way in `testlapack` to check orthogonality of matrices. I'll remove the others and change the test to use this one.

kortschak replied:
ok

vladimir-ch commented on lapack/testlapack/dtrexc.go:
```diff
@@ -205,16 +199,126 @@ func testDtrexc(t *testing.T, impl Dtrexcer, compq lapack.UpdateSchurComp, tmat
+	return dlansy(lapack.MaxColumnSum, blas.Upper, work.N, work.Data, work.Stride)
+}
+
+func dlansy(norm lapack.MatrixNorm, uplo blas.Uplo, n int, a []float64, lda int) float64 {
```
As with `dlange` I'm not sure if it's better to add these methods to the tested interface or use these duplicated but local, independent implementations. I slightly prefer this but you may have a different opinion.

kortschak replied:
I'm fine with this.

kortschak commented on lapack/testlapack/general.go:
```diff
@@ -1158,29 +1158,51 @@ func extract2x2Block(t []float64, ldt int) (a, b, c, d float64) {
 // isSchurCanonical returns whether the 2×2 matrix [a b; c d] is in Schur
 // canonical form.
 func isSchurCanonical(a, b, c, d float64) bool {
-	return c == 0 || (a == d && math.Signbit(b) != math.Signbit(c))
```
I actually find the original easier to read.

`return c == 0 || (b != 0 && a == d && math.Signbit(b) != math.Signbit(c))`

vladimir-ch replied:
You're right. Done.

vladimir-ch commented on lapack/testlapack/general.go:
```diff
@@ -131,8 +131,8 @@ func randomHessenberg(n, stride int, rnd *rand.Rand) blas64.General {
 // elements of opposite sign.
-func randomSchurCanonical(n, stride int, rnd *rand.Rand) blas64.General {
-	t := randomGeneral(n, n, stride, rnd)
+func randomSchurCanonical(n, stride int, rnd *rand.Rand) (t blas64.General, wr, wi []float64) {
```
Here I'm following the lapack namin for these slices/arrays but I don't know why lapack calls them like this.

vladimir-ch commented:
I tested with netlib and I get only errors for unexpected eigenvalues of 2x2 blocks, not for the matrix Q. This confirms that there will be a bug somewhere in the updates to Q but yesterday I didn't see any mistake in Dtrexc quickly. I'll check again and also the call dependencies.

kortschak commented:
I tested with netlib earlier (a29e678437e40d1d2c47ed1171b760d192089c5a) and got no errors (and now again with 520dc6057072ee5d1d9e40dec6bfcabb29f13884). When I checkout 04cb64a26ccf96b185ca61788c0f6189abe35abc, and run the netlib test I get many "unexpected {first,second} eigenvalue of 2×2 block".
